README.md
# External links

[![Build Status](https://travis-ci.org/silverstripe-labs/silverstripe-externallinks.svg?branch=master)](https://travis-ci.org/silverstripe-labs/silverstripe-externallinks)

## Introduction

The external links module is a task and ModelAdmin to track and to report on broken external links.

## Maintainer Contact

 * Damian Mooyman (@tractorcow) <damian@silverstripe.com>

## Requirements

	* SilverStripe 3.1 +

## Features

* Add external links to broken links reports
* Add a task to track external broken links

See the [changelog](CHANGELOG.md) for version history.

## Installation

 1. If you have composer you can use `composer require silverstripe/externallinks:*`. Otherwise, 
    download the module from GitHub and extract to the 'externallinks' folder. Place this directory
    in your sites root directory. This is the one with framework and cms in it.
 2. Run in your browser - `/dev/build` to rebuild the database.
 3. Run the following task *http://path.to.silverstripe/dev/tasks/CheckExternalLinks* to check for
    broken external links

## Report ##

A new report is added called 'External Broken links report'. When viewing this report, a user may press
the "Create new report" button which will trigger an ajax request to initiate a report run.

In this initial ajax request this module will do one of two things, depending on which modules are included:

* If the queuedjobs module is installed, a new queued job will be initiated. The queuedjobs module will then
  manage the progress of the task.
* If the queuedjobs module is absent, then the controller will fallback to running a buildtask in the background.
  This is less robust, as a failure or error during this process will abort the run.

In either case, the background task will loop over every page in the system, inspecting all external urls and
checking the status code returned by requesting each one. If a URL returns a response code that is considered
"broken" (defined as < 200 or > 302) then the `ss-broken` css class will be assigned to that url, and 
a line item will be added to the report. If a previously broken link has been corrected or fixed, then
this class is removed.

In the actual report generated the user can click on any broken link item to either view the link in their browser,
or edit the containing page in the CMS.

While a report is running the current status of this report will be displayed on the report details page, along
with the status. The user may leave this page and return to it later to view the ongoing status of this report.

Any subsequent report may not be generated until a prior report has completed.

## Dev task ##

Run the following task *http://path.to.silverstripe/dev/tasks/CheckExternalLinks* to check your site for external
broken links.

## Queued job ##

If you have the queuedjobs module installed you can set the task to be run every so ofter
Add the following yml config to config.yml in mysite/_config have the the task run once every day (86400 seconds)

    CheckExternalLinks:
      Delay: 86400

## Whitelisting codes ##

If you want to ignore or whitelist certain http codes this can be setup via IgnoreCodes in the config.yml
file in mysite/_config

    CheckExternalLinks:
      Delay: 60
      IgnoreCodes:
        - 401
        - 403
        - 501


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.3]

* Changelog added.
* Update translations.


code/tasks/CheckExternalLinks.php
<?php

class CheckExternalLinks extends BuildTask {
	public $limit = 10;

	protected $title = 'Checking broken External links in the SiteTree';

	protected $description = 'A task that records external broken links in the SiteTree';

	protected $enabled = true;

	private $completedPages;
	private $totalPages;

	function run($request) {
		$this->runLinksCheck($this->limit);
	}

	/**
	 * Runs the links checker and returns the track used
	 *
	 * @param int $limit Limit to number of pages to run
	 * @return BrokenExternalPageTrackStatus
	 */
	public function runLinksCheck($limit) {
		$track = CheckExternalLinks::getLatestTrack();

		// if the script has already been started
		if ($track && $track->Status == 'Running') {
			$batch = BrokenExternalPageTrack::get()
				->filter(array(
					'TrackID' => $track->ID,
					'Processed' => 0
				))->limit($limit)->column('PageID');
			$pages = Versioned::get_by_stage('SiteTree', 'Stage')
				->filter('ID', $batch)
				->limit($limit);
			$this->updateJobInfo('Fetching pages to check');
			if ($track->CompletedPages == $track->TotalPages) {
				$track->Status = 'Completed';
				$track->write();
				$this->updateJobInfo('Setting to completed');
			}
		// if the script is to be started
		} else {
			$pages = Versioned::get_by_stage('SiteTree', 'Stage')->column('ID');
			$noPages = count($pages);

			$track = BrokenExternalPageTrackStatus::create();
			$track->TotalPages = $noPages;
			$track->write();
			$this->updateJobInfo('Creating new tracking object');

			foreach ($pages as $page) {
				$trackPage = BrokenExternalPageTrack::create();
				$trackPage->PageID = $page;
				$trackPage->TrackID = $track->ID;
				$trackPage->write();
			}

			$batch = BrokenExternalPageTrack::get()
				->filter(array(
					'TrackID' => $track->ID
				))->limit($limit)->column('PageID');

			$pages = Versioned::get_by_stage('SiteTree', 'Stage')
				->filter('ID', $batch);
		}
		$trackID = $track->ID;
		foreach ($pages as $page) {
			++$this->totalPages;

			if ($track->ID) {
				$trackPage = BrokenExternalPageTrack::get()
					->filter(array(
						'PageID' => $page->ID,
						'TrackID' => $track->ID
					))->first();
				$trackPage->Processed = 1;
				$trackPage->write();
			}

			$htmlValue = Injector::inst()->create('HTMLValue', $page->Content);
			if (!$htmlValue->isValid()) {
				continue;
			}

			// Populate link tracking for internal links & links to asset files.
			if($links = $htmlValue->getElementsByTagName('a')) foreach($links as $link) {
				$class = $link->getAttribute('class');
				$pos = stripos($class, 'ss-broken');
				if ($pos !== false && $page->HasBrokenLink == 1) continue;

				$href = Director::makeRelative($link->getAttribute('href'));
				if ($href == 'admin/') continue;

				// ignore SiteTree, anchor and assets links as they will be caught
				// by SiteTreeLinkTracking
				if(preg_match('/\[(file_link|sitetree_link),id=([0-9]+)\]/i', $href, $matches)) {
					continue;
				} else if (isset($href[0]) && $href[0] == '#') {
					continue;
				} else if(substr($href, 0, strlen(ASSETS_DIR) + 1) == ASSETS_DIR.'/') {
					continue;
				}

				if($href && function_exists('curl_init')) {
					$handle = curl_init($href);
					curl_setopt($handle, CURLOPT_RETURNTRANSFER, TRUE);
					curl_setopt($handle, CURLOPT_CONNECTTIMEOUT, 5);
					curl_setopt($handle, CURLOPT_TIMEOUT, 10);
					$response = curl_exec($handle);
					$httpCode = curl_getinfo($handle, CURLINFO_HTTP_CODE);  
					curl_close($handle);
					// do we have any whitelisted codes
					$ignoreCodes = Config::inst()->get('CheckExternalLinks', 'IgnoreCodes');
					// if the code is whitelisted set it to 200
					$httpCode  = (is_array($ignoreCodes) && in_array($httpCode, $ignoreCodes)) ?
						200 : $httpCode;

					// ignore empty hrefs and internal links
					if (($httpCode < 200 || $httpCode > 302) || ($href == '' || $href[0] == '/')) {
						$brokenLink = new BrokenExternalLink();
						$brokenLink->PageID = $page->ID;
						$brokenLink->Link = $href;
						$brokenLink->HTTPCode = $httpCode;
						$brokenLink->write();

						// set the broken link class
						$class = ($class && stripos($class, 'ss-broken')) ?
							$class . ' ss-broken' : 'ss-broken';
						$link->setAttribute('class', ($class ? $class : 'ss-broken'));
						$htmlValue->__call('saveHTML', array());

						$page->Content = $htmlValue->getContent();
						$page->owner->write();

						if (!$page->HasBrokenLink) {

							// bypass the ORM as syncLinkTracking does not allow you
							// to update HasBrokenLink to true
							$query = "UPDATE \"SiteTree\" SET \"HasBrokenLink\" = 1 ";
							$query .= "WHERE \"ID\" = " . (int)$page->ID;   
							$result = DB::query($query);
							if (!$result) {
								$this->debugMessage('Error updating HasBrokenLink');
							}
						}

					}
				}
			}
			++$this->completedPages;
		}

		// run this outside the foreach loop to stop it locking DB rows
		$this->updateJobInfo('Updating completed pages');
		$this->updateCompletedPages($trackID);

		// do we need to carry on running the job
		$track = $this->getLatestTrack();
		if ($track->CompletedPages >= $track->TotalPages) {
			$track->Status = 'Completed';
			$track->write();

			// clear any old previous data
			$rows = BrokenExternalPageTrack::get()
				->exclude('TrackID', $track->ID);
			foreach ($rows as $row) {
				$row->delete();
			}
			return $track;
		}

			// if running via the queued job module return to the queued job after each iteration
		if ($limit == 1) {
			return $track;
			} else {
				$this->updateJobInfo("Running next batch {$track->CompletedPages}/{$track->TotalPages}");
			return $this->runLinksCheck($limit);
		}
	}

	public static function getLatestTrack() {
		$track = BrokenExternalPageTrackStatus::get()->sort('ID', 'DESC')->first();
		if (!$track || !$track->exists()) return null;
		return $track;
	}

	public static function getLatestTrackID() {
		$track = CheckExternalLinks::getLatestTrack();
		if (!$track || !$track->exists()) return null;
		return $track->ID;
	}

	public static function getLatestTrackStatus() {
		$track = CheckExternalLinks::getLatestTrack();
		if (!$track || !$track->exists()) return null;
		return $track->Status;
	}

	private function updateCompletedPages($trackID = 0) {
		$noPages = BrokenExternalPageTrack::get()
			->filter(array('TrackID' => $trackID, 'Processed' => 1))->count();
		$track = $this->getLatestTrack($trackID);
		$track->CompletedPages = $noPages;
		$track->write();
		return $noPages;
	}

	private function updateJobInfo($message) {
		$track = CheckExternalLinks::getLatestTrack();
		if (!$track || !$track->exists()) return null;
		$track->JobInfo = $message;
		$track->write();
	}
}
